### DISCUSSION TITLE
Operate over already allocated memory

### DISCUSSION THREAD

**Discussion by DanielLazarHTDM** [nlohmann/json#3839] [Ideas]
Hi,

I have a question about maybe improving this library. I was wondering if it is possible to operate over already allocated memory.
My motivation is because of inter process communication. I am using https://github.com/eclipse-iceoryx/iceoryx for it which is based on shared memory between processes and zero copy approach. I would like to share JSON between two processes. 

With current implementation I would need to:
1. Create json object
2. Fill it by this great API
3. Serialize it by BSON
4. COPY it to shared memory (iceoryx loan, publisher)
5. Parse it from BSON received shared memory in other process (iceoryx subscriber)
6. Use JSON in another process

I was thinking about it and there are two copy which are not necessary from my point of view. Suggested way if it is possible.
1. Statically allocated memory of e.g. of size 4096B (of course it will be up to developer to count it correctly)
2. Json object which will take this pointer
3. Filling object by this great public API
4. Send buffer using shared memory into another process
5. The second process will just again wrap object (like cast pointer into type) and use it directly.

I know there are arrays. Unknown size of fields etc. I am just asking if it has technicall solution. Maybe we can help by some static map of fields data types to have it "serialized" directly during operating with final buffer.  

It would be great to achieve this since another process in iox-environment will have access to same shared memory so pointer addresses will fit and simply works without copy ;-)

Boost seems to have something like that. https://www.boost.org/doc/libs/1_80_0/libs/json/doc/html/json/quick_look.html

Let me know, thank you Dan

**Comment by falbrechtskirchinger** [comment]
The library is allocator-aware, so that'd be the place to start.

**Reply by DanielLazarHTDM** [reply]
I found https://json.nlohmann.me/api/basic_json/get_allocator/. But no idea how to use already allocated memory here. Sorry not so good with this dynamic std::allocators. Do you think you can create a hint for me just with few lines of code?

**Reply by falbrechtskirchinger** [reply]
As far as the library is concerned, you only need to supply your allocator type to the `basic_lson` template:
```cpp
template<
    template<typename U, typename V, typename... Args> class ObjectType = std::map,
    template<typename U, typename... Args> class ArrayType = std::vector,
    class StringType = std::string,
    class BooleanType = bool,
    class NumberIntegerType = std::int64_t,
    class NumberUnsignedType = std::uint64_t,
    class NumberFloatType = double,
    template<typename U> class AllocatorType = std::allocator,                             <-----------
    template<typename T, typename SFINAE = void> class JSONSerializer = adl_serializer,
    class BinaryType = std::vector<std::uint8_t,
    class CustomBaseClass = void>
>
class basic_json;
```

And I'm sure a quick search will turn up any number of tutorials/articles on the subject of custom allocators. It's unfortunately not something that can be done in a handful of lines.

Maybe you can find an implementation of a custom allocator that suits your needs. (Boost provides some, for example.)
